### PR TITLE
Adding in a way of plugin defining route to non-MVC API Extensions

### PR DESCRIPTION
**Pull request by wilsonge**
Version 1 of getting a response. With this you now get a URL of http://localhost/~george/joomla-cms/api/index.php/article giving:

`{"links":{"self":"http:\/\/localhost\/~george\/joomla-cms\/api\/index.php\/article"}}`

Next step is to have the plugin inject a mapping object which runs though and maps the object to the response

### CODE REVIEW

File: api/components/com_content/dispatcher.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getController($name, $client = null, $config = array())
+	{
+		// Set up the namespace
+		$namespace = rtrim($this->namespace, '\\') . '\\';
+
+		// Set up the client
+		$client = 'Site';
```
Comment (wilsonge) [inline api/components/com_content/dispatcher.php]: Only difference here between this and the site dispatcher is forcing the client to site for the controller creation

File: libraries/src/CMS/Application/ApiApplication.php [hunk 2]
```diff
@@ -267,4 +266,19 @@ public function dispatch($component = null)
...
+	 * @param   string  $name     The name of the application/client.
+	 * @param   array   $options  An optional associative array of configuration settings.
+	 *
+	 * @return  AbstractMenu|null
+	 *
+	 * @since   3.2
+	 */
+	public function getMenu($name = null, $options = array())
+	{
+		return null;
```
Comment (wilsonge) [inline libraries/src/CMS/Application/ApiApplication.php]: Not sure what to do about this. This was a hack for working around a menu item being required at: https://github.com/joomla-projects/gsoc17_webservices/blob/4.0-dev/libraries/src/CMS/Controller/Controller.php#L750

File: libraries/src/CMS/Controller/Controller.php [hunk 3]
```diff
@@ -852,6 +852,12 @@ public function getView($name = '', $type = '', $prefix = '', $config = array())
 			else
 			{
 				$prefix = $this->app->getName();
+
+				if ($prefix === 'api')
```
Comment (wilsonge) [inline libraries/src/CMS/Controller/Controller.php]: Hack around to use the right model - this should be moved to the MvcFactory

File: libraries/src/CMS/Document/Factory.php [hunk 4]
```diff
@@ -32,6 +32,11 @@ public function createDocument($type = 'html', array $attributes = array())
 		$type  = preg_replace('/[^A-Z0-9_\.-]/i', '', $type);
 		$ntype = null;
 
+		if ($type === 'applicationvnd.apijson')
```
Comment (wilsonge) [inline libraries/src/CMS/Document/Factory.php]: Ensure's the right JDocument class is loaded (from the JInput format)

File: libraries/joomla/document/api/jsonapi.php [hunk 5]
```diff
@@ -26,7 +26,7 @@ class JDocumentApiJsonapi extends JDocumentJson implements JsonSerializable
 	 * @var    Document
 	 * @since  __DEPLOY_VERSION__
 	 */
-	protected $document;
+	public $document;
```
Comment (wilsonge) [inline libraries/joomla/document/api/jsonapi.php]: Do we want to keep this protected and just have public wrapping methods for everything? Or do we want to make it public?

File: libraries/joomla/document/jsonapi.php [hunk 6]
```diff
@@ -26,7 +26,7 @@ class JDocumentJsonapi extends JDocumentJson implements JsonSerializable
 	 * @var    Document
 	 * @since  __DEPLOY_VERSION__
 	 */
-	protected $document;
+	public $document;
```
Comment (wilsonge) [inline libraries/joomla/document/jsonapi.php]: We need to decide whether to create wrapper methods for everything (like the `addLink` method I'm using in com_content view, or make this public and delete any proxies to that method)

File: libraries/joomla/document/jsonapi.php [hunk 7]
```diff
@@ -41,6 +41,7 @@ public function __construct($options = array())
 
 		// Set mime type to JSON-API
 		$this->_mime = 'application/vnd.api+json';
+		$this->_type = 'jsonapi';
```
Comment (wilsonge) [inline libraries/joomla/document/jsonapi.php]: We need to decide on whether to make this type explicit or not as it defines what the View Format is (i.e. Should component's be limited to a single JSON view or should be have json+ld and jsonapi and json views all available?

File: api/components/com_content/dispatcher.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	 *
+	 * @return  Controller
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function getController($name, $client = null, $config = array())
+	{
+		if ($client === null || strtolower($client) === 'api')
+		{
+			$client = 'Site';
```
Comment (wilsonge) [inline api/components/com_content/dispatcher.php]: Hack to force using the site controller for com_content to test things working

File: libraries/src/CMS/Application/ApiApplication.php (hunk 2 above)
Comment (wilsonge) [reply libraries/src/CMS/Application/ApiApplication.php]: Pathway also relies on the menu existing. Unsure whether to make that return `null` too

**Comment by fastslack** [comment]
@wilsonge I have some problems patching this PR

```
$ patch -p1 <36.patch 
The next patch would create the file api/components/com_content/dispatcher.php,
which already exists!  Assume -R? [n] 
Apply anyway? [n] 
Skipping patch.
1 out of 1 hunk ignored
The next patch would create the file components/com_content/View/Article/Json.php,
which already exists!  Assume -R? [n] 
Apply anyway? [n] 
Skipping patch.
1 out of 1 hunk ignored
can't find file to patch at input line 172
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/libraries/joomla/document/api/jsonapi.php b/libraries/joomla/document/api/jsonapi.php
|index 101b4ae8c4..f06b1ebc3f 100644
|--- a/libraries/joomla/document/api/jsonapi.php
|+++ b/libraries/joomla/document/api/jsonapi.php
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
1 out of 1 hunk ignored
patching file libraries/src/CMS/Application/ApiApplication.php
Hunk #2 FAILED at 119.
Hunk #3 succeeded at 297 (offset 29 lines).
1 out of 3 hunks FAILED -- saving rejects to file libraries/src/CMS/Application/ApiApplication.php.rej
patching file libraries/src/CMS/Controller/Controller.php
patching file libraries/src/CMS/Document/Factory.php
patching file components/com_content/View/Article/Jsonapi.php (renamed from components/com_content/View/Article/Json.php)
patching file libraries/joomla/document/jsonapi.php (already renamed from libraries/joomla/document/api/jsonapi.php)
Reversed (or previously applied) patch detected!  Assume -R? [n] 
Apply anyway? [n] 
Skipping patch.
2 out of 2 hunks ignored -- saving rejects to file libraries/joomla/document/jsonapi.php.rej
patching file libraries/src/CMS/Application/ApiApplication.php
Reversed (or previously applied) patch detected!  Assume -R? [n] 
Apply anyway? [n] 
Skipping patch.
3 out of 3 hunks ignored -- saving rejects to file libraries/src/CMS/Application/ApiApplication.php.rej
patching file libraries/src/CMS/Controller/Controller.php
patching file api/components/com_content/View/Article/Jsonapi.php (renamed from components/com_content/View/Article/Jsonapi.php)
patching file libraries/bootstrap.php
Reversed (or previously applied) patch detected!  Assume -R? [n] 
Apply anyway? [n] 
Skipping patch.
1 out of 1 hunk ignored -- saving rejects to file libraries/bootstrap.php.rej
patching file libraries/src/CMS/Controller/Controller.php
patching file libraries/src/CMS/Document/Factory.php
patching file api/components/com_content/dispatcher.php
```

But after manually patch, it works

File: libraries/joomla/document/jsonapi.php (hunk 6 above)
Comment (fastslack) [reply libraries/joomla/document/jsonapi.php]: I think there is no problem to leave document public

File: libraries/joomla/document/jsonapi.php (hunk 7 above)
Comment (fastslack) [inline libraries/joomla/document/jsonapi.php]: For first instancese we can use fixed type but for future would be awesome to have some switch for differents json types

**Comment by wilsonge** [comment]
It looks like you're trying to go patching patch. I've fixed some of the stuff here into the main lib_api branch as I fixed it. Like there's no changes to the libraries/src/CMS/Controller/Controller.php file here anymore :/

File: libraries/joomla/document/jsonapi.php (hunk 6 above)
Comment (wilsonge) [reply libraries/joomla/document/jsonapi.php]: I think it depends how much you want to rely on the implementation. For example we can add wrapper interfaces on some of the library interfaces and swap it out if we ever need to. If you make this public we're totally dependent on the library for life

File: libraries/src/CMS/Mvc/Factory/MvcFactory.php [hunk 9]
```diff
@@ -62,6 +62,11 @@ public function __construct($namespace, CMSApplicationInterface $application)
 	 */
 	public function createModel($name, $prefix = '', array $config = array())
 	{
+		if (empty($prefix) && $this->application->isClient('api'))
```
Comment (wilsonge) [inline libraries/src/CMS/Mvc/Factory/MvcFactory.php]: Dirty hack to ensure that models are correctly loaded. We should use our own ApiFactory

**Pull request merged**